### DISCUSSION TITLE
[REQUEST] Resumability

### DISCUSSION THREAD

**Discussion by AmeerArsala** [withastro/roadmap#1040] [Proposal]
### Body

# Summary

I would like the feature of Resumability (what QwikJS does) in Astro.

# Background & Motivation

I love Astro and can't imagine my life without it. However, resumability is one feature that I feel would make it shoot past NextJS more than it already has. Whereas with partial hydration it selectively injects JS where the dev specifies, resumability delays JS execution as much as possible.

This would be particularly useful in cross-platform development. I am building an application with Astro and Tauri that will deploy to Web, Desktop (Windows, Mac, Linux), and Mobile (iOS and Android) all with 1 codebase. However, the way I do this sometimes is by selectively showing and hiding components (e.g. desktop components, mobile components, etc.) with tailwindcss. However, those components oftentimes will execute JS even when hidden (such as a desktop component executing on the mobile platform). 

By declaring certain components resumable, this automatically occurs 

# Goals

- Be able to declare certain components resumable

# Example

A `client:resume` directive as in `<MySvelteComponent client:resume/>`

**Comment by ematipico** [comment]
I'm not very familiar with the concept of resumability. Your proposal mentions it, but it does it as if the reader already knows the meaning and how it works. Would you mind adding an explanation for people like me that don't know the concept very well?

**Reply by adaliszk** [reply]
The https://qwik.dev/docs/concepts/resumable page has an excellent summary of it. However, in a nutshell, resumability is the next level after hydration where instead of assuming that you need to rebuild the component tree, the state, and re-attach listeners, you encode everything into the HTML in a way where you can on the fly "hydrate" every dynamic portion (like variables in the template) when the users interact with the components.

Fundamentally, it recognizes that the output of a complex application does not need to reflect that application, and what matters is what changes and what triggers that change. Any mutation must only be applied when a user triggers an interaction; until that point, everything can be entirely static.

**Comment by wmertens** [comment]
Note that you can already use Qwik in Astro. Resumability is not trivial to add to frameworks, the entire SSR stack needs to be written with it in mind, and you need to make design decisions that may conflict with how frameworks work right now.

**Reply by adaliszk** [reply]
Yeah, I suppose what would be an improvement there is to see (like the same level as any other frameworks, not just a single word mention) Qwik on the https://docs.astro.build/en/guides/framework-components/ page. That way, people would find it without needing to search explicitly or stumble upon it on the Qwik Docs.

**Reply by AmeerArsala** [reply]
I understand that you can, but I want to use resumable Svelte components (for example) so I can use Svelte libraries with it. Qwik does not yet have an amazing ecosystem, so it is very much lacking in usability, but the use-case I described above would be perfect for resumibility, and would likely make Astro (+ Tauri) the undisputed leader in cross-platform software development.

**Reply by wmertens** [reply]
Like I said, each framework would have to implement resumability itself. Qwik doesn't even require anything on the Astro side. 

You can't force a framework to be resumable.

**Reply by AmeerArsala** [reply]
I can understand that applying to React/Preact which tries to be straight up raw JS(X) hooks, but all the other major frameworks are using compilers (and React is starting to adopt it too) that compile to just JS. So if they all end up compiling to pure JS anyway, then I'm sure there is a way for resumability to be added as that would be the common denominator.

**Reply by adaliszk** [reply]
Resumability relies on compilation steps, where the interaction and state boundaries are detected, and those entrypoints are extracted and encoded in the final output. Resumability in Qwik is added through that specific way of compilation and not through transpiling pure JS outputs.

Maybe, one day, `qwikify$` could be adapted to other frameworks. It likely will require collaboration with the various frameworks to expose hooks both in their JSX and compilation steps, and the Qwik team would need to spend a lot of time on it, which they do not have in the foreseeable future.

Regardless, this is a topic for framework authors and has nothing to do with Astro. Once the frameworks expose anything other than hydration, Astro could provide that. Still, until that happens, Astro could only change its documentation and its very own component solution, where the latter is already HTML-first and very lightweight; adding resumable interaction and state may not significantly improve any metric.

**Reply by wmertens** [reply]
BTW just to clarify, qwikify$ doesn't make React resumable, it has to hydrate the React part

**Comment by adaliszk** [comment]
Unfortunately, reusability is not something you can inject into all frameworks; they all would need to implement the paradigm or collaborate with the Qwik team to expose hooks. Essentially rewriting significant portions of their entire framework or complication step. However, wouldn't `client:visible` work in your case? That only hydrates the components when they become visible, which should work with your current solution of hiding and showing elements with classes.

**Reply by adaliszk** [reply]
Other than that, you can actually use Qwik as well for the particular components that need resumability.

**Comment by apatel369** [comment]
You can achieve partial resumability by hydrating framework components on click using custom client directive. https://docs.astro.build/en/reference/integrations-reference/#addclientdirective-option
